### DISCUSSION TITLE
Better APIs and Modules to Implement Security, Authentication and Authorisation

### DISCUSSION THREAD

**Discussion by sirinath** [activej/activej#191] [Ideas]
For web-facing deployments security is paramount, hence it is a requirement to protect the deployments against hackers. For this is it possible to implement security, authentication and authorisation which might include:
- OAuth (Facebook, Twitter, Google, etc)
- SAML
- CAS 
- OpenID Connect
- OpenID
- JWT
- Kerberos
- etc.
with authorization mechanisms:
- Roles/permissions 
- Anonymous/remember-me/(fully) authenticated 
- CORS 
- CSRF 
- HTTP Security Headers
- etc.
either internally or integrating with an existing framework.

This issue was initially raised in a different form at DataKernal: https://github.com/softindex/datakernel/issues/20

**Comment by eduard-vasinskyi** [comment]
Hello, @sirinath 
ActiveJ framework is designed in such way to have no dependencies on external protocols. Rather, it provides all the necessary tools for developers to implement custom protocols for specific business needs.

There is a [tutorial](https://activej.io/tutorials/authorization-tutorial.html) on how to implement simplistic authorization service using ActiveJ framework. It is a good start if you want to implement some of the existing authorization protocols. You can use any authorization engine on top of ActiveJ's HTTP server and client. If you have any questions or troubles regarding implementing authorization, please feel free to ask.

**Comment by sirinath** [comment]
For the time bean, I am using PAC4J but I am not sure if there can be a better, faster and lighter GC-less implementation. But if ActiveJ implements its own module I would glad to use this than have another dependency.

**Comment by tr7zw** [comment]
It would be really helpful to have an PAC4J example/premade lib like for [spring](https://github.com/pac4j/spring-webmvc-pac4j)/[vertx](https://github.com/pac4j/vertx-pac4j)/etc.

**Comment by sirinath** [comment]
Also [SuperTokens ](https://github.com/supertokens/supertokens-core) might be of interest.
